File: Data Structures & Algorithms/merge-intervals/submission-3.swift
class Solution {
    func merge(_ intervals: [[Int]]) -> [[Int]] {
        guard !intervals.isEmpty else { return [] }

        // O(n) - Pre-sort list so we can use an expanding strat
        let sIntervals: [[Int]] = intervals.sorted{$0[0] < $1[0]}

        // O(n) - Merge Each Interval into the next
        var retInts: [[Int]] = []
        var a = sIntervals[0]
        for b in sIntervals[1...] {
            if b[0] <= a[1] {
                if b[1] >= a[1] {
                    // b start NOT end is within a
                    a[1] = b[1]
                }
            } else {
                retInts.append(a)
                a = b
            }
        }
        retInts.append(a)

        return retInts
    }
}

File: Data Structures & Algorithms/meeting-schedule/submission-3.ts
/**
 * Definition of Interval:
 * class Interval  {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

class Solution {
    /**
     * @param {Interval[]} intervals
     * @returns {boolean}
     */
    canAttendMeetings(intervals: Interval[]): boolean {
        if (intervals.length == 0) { return true }
        intervals = intervals.sort((a,b) => a.start - b.start )
        let prevEnd = intervals[0].end

        for (let interval of intervals.slice(1)) {
            if (prevEnd > interval.start) {
                return false
            }
            prevEnd = interval.end
        }
        return true
    }
}
